apps/communications/contacts/test/unit/views/form/form_ui_test.js

```diff
@@ -0,0 +1,521 @@
+        return element.getAttribute('name').startsWith('comment');
+      });
+
+      // After cleaning both entries, we must have no phone number
+      // input in the form
+      assert.equal(inputs.length, 0);
+    });
+  });
+
+
+  suite('Date', function() {
+    test(' > Save button must be disabled before adding date', function() {
+      var saveButton = document.querySelector('#save-button');
+      assert.equal(saveButton.getAttribute('disabled'), 'disabled');
+    });
+
+    test(' > Add Date manually. Must be "birthday"', function() {
+      document.getElementById('add-new-date').click();
+      var inputs =
+        document.querySelectorAll('input[type="date"]:not([id="date_#i#"])');
+      assert.equal(inputs[0].dataset.field, 'birthday');
+      assert.equal(inputs.length, 1);
+    });
+
+    test(' > Add a second Date manually. Must be "Anniversary"', function() {
+      document.getElementById('add-new-date').click();
+      var inputs =
+        document.querySelectorAll('input[type="date"]:not([id="date_#i#"])');
+      assert.equal(inputs[1].dataset.field, 'anniversary');
+      assert.equal(inputs.length, 2);
+    });
+
+    test(' > We can not add more Dates! Max is 2.', function() {
+      // Click several times
+      document.getElementById('add-new-date').click();
+      document.getElementById('add-new-date').click();
+      document.getElementById('add-new-date').click();
+      document.getElementById('add-new-date').click();
+
+      // Inputs are not growing
+      var inputs =
+        document.querySelectorAll('input[type="date"]:not([id="date_#i#"])');
+      assert.equal(inputs.length, 2);
+
+      // Button for adding must be disabled
+      assert.isTrue(document.querySelector('#add-new-date').disabled);
+    });
+
+
+    test(' > Remove Dates', function() {
+      var deleteButtons =
+        document.querySelectorAll('button[data-action="delete"]');
+
+      var touchstart = 'ontouchstart' in window ? 'touchstart' : 'mousedown';
+      var customEvent = new CustomEvent(
+        touchstart,
+        {
+          bubbles: true,
+          cancelable: true
+        }
+      );
+      for (var i = 0; i < deleteButtons.length; i++) {
+        deleteButtons[i].dispatchEvent(customEvent);
+      }
+
+      var inputs =
+        document.querySelectorAll('input[type="date"]:not([id="date_#i#"])');
+
+      // After cleaning both entries, we must have no dates
+      // input in the form
+      assert.equal(inputs.length, 0);
+    });
+  });
+
+  suite('Save button', function() {
+    var contact = {
+      tel: ['123456789']
+    };
+
+    test(' > must dispatch an event when clicked', function(done) {
+      FormUI.render(contact);
+
+      window.addEventListener('save-contact', function() {
+        done();
+      });
+
+      var clickEvent = new CustomEvent('click', {'detail': contact});
+      document.querySelector('#save-button').dispatchEvent(clickEvent);
+    });
+  });
+});
```

Comment:
Love the way we have suites for each of the different components of a form. Amazing job.